lite2/client.go

```diff
@@ -916,25 +879,35 @@ func (c *Client) backwards(toHeight int64, fromHeader *types.SignedHeader, now t
 func (c *Client) compareNewHeaderWithWitnesses(h *types.SignedHeader) error {
 	c.providerMutex.Lock()
 	defer c.providerMutex.Unlock()
-	// 0. Check witnesses exist
-	if len(c.witnesses) == 0 {
-		return errors.New("could not find any witnesses")
-	}
-
-	matchedHeader := false
 
+	// 1. Make sure AT LEAST ONE witness returns the same header.
+	headerMatched := false
+	witnessesToRemove := make([]int, 0)
 	for attempt := uint16(1); attempt <= c.maxRetryAttempts; attempt++ {
-		// 1. Loop through all witnesses.
-		for _, witness := range c.witnesses {
+		if len(c.witnesses) == 0 {
+			return errors.New("could not find any witnesses. please reset the light client")
+		}
 
-			// 2. Fetch the header.
+		for i, witness := range c.witnesses {
 			altH, err := witness.SignedHeader(h.Height)
 			if err != nil {
-				c.logger.Info("No Response from witness ", "witness", witness)
+				c.logger.Error("Failed to get a header from witness", "height", h.Height, "witness", witness)
+				continue
+			}
+
+			if err = altH.ValidateBasic(c.chainID); err != nil {
+				c.logger.Error("Witness sent us incorrect header", "err", err, "witness", witness)
+				witnessesToRemove = append(witnessesToRemove, i)
+				continue
+			}
+
+			if err = c.trustedNextVals.VerifyCommitTrusting(c.chainID, altH.Commit.BlockID,
+				altH.Height, altH.Commit, c.trustLevel); err != nil {
+				c.logger.Error("Witness sent us incorrect header", "err", err, "witness", witness)
+				witnessesToRemove = append(witnessesToRemove, i)
```

Comment:
What happens if we notice double voting. Shouldn't we deal with that or should we just ignore it

Reply:
actually why is this necessary. If I understand we are checking that there is a one third overlap between the next validator set at h + 1 and the witnesses current block at h.

Reply:
https://github.com/tendermint/spec/pull/71#discussion_r374213456

Reply:
okay no you are right because we haven't update trustedNextVals yet